text/3680-use.md

```diff
@@ -0,0 +1,463 @@
+- `Result<T, E>` where `T: Use` and `E: Use`
+
+Some notable types that implement `Clone` but should *not* implement `Use`:
+arrays, `String`, `Box`, `Vec`, `HashMap`, and `BTreeMap`.
+
+We may want to add a clippy or rustc lint (e.g. `expensive_use`) for
+implementations of `Use` on an excessively large type, or a type whose `clone`
+implementation seems to be obviously breaking the intended constraints on the
+`Use` trait.  Such a lint would be best-effort only, and could always be marked
+as `allow` by a crate, but could help to discourage such implementations.
+
+We may want to add a clippy or rustc lint for calls to `.clone()` that could
+use `.use` instead. This would help the remaining calls to `.clone()` stand out
+as "expensive" clones. (Such a lint would need to take MSRV into account before
+making such a suggestion; clippy already has such a mechanism and rustc may
+gain one in the future.)
+
+We may want to add a `derive(Use)`, which automatically adds `Use` bounds on
+fields just as `derive(Clone)` adds `Clone` bounds on fields.
+
+## The implementation and optimization of `.use`
+
+An expression `x.use`, where `x` has type `T`, requires that `T: Use`. However,
+`x.use` does not always invoke `Clone::clone(x)`; in some cases the compiler
+can optimize away a use.
+
+If `x` is owned and statically known to be dead, the compiler will move `x`
+rather than using it. This allows functions to write `x.use` without concern
+for whether it's the last usage of `x` (e.g. when passing `x.use` to a series
+of functions). Much like a trailing comma, this allows every usage to be
+symmetric, making it easy to compare uses or add more uses afterwards. (This
+optimization also means we should generally not lint on a final use of `x.use`,
+such as we currently do with the clippy lint `redundant_clone`.)
+
+If `x` is not statically known to be dead, but *all* of the following
+conditions are met, the compiler *may* elide an `x.use` and use `&x` instead:
+- The compiler can statically see that `x` outlives the result of `x.use`,
+- The compiler can statically see that the result of `x.use` is only accessed
+  via shared reference (including methods with `&self`, dereferences via
+  `Deref`, other invocations of `.use`, `use ||` closures, or `async use`
+  blocks). Effectively, these conditions mean that the user could theoretically
+  have refactored the code to use `&x` rather than `x.use`.
+
+An example of these elisions:
+
+```rust
+fn f(obj: Arc<Object>) {
+    g(obj.use); // `g` takes `Arc<Object>`
+    h(use || obj.method()); // `Object::method` here takes `&self`
+}
+
+fn main() {
+    let obj = Arc::new(Object::new());
+    f(obj.use);
+    f(obj.use);
+    f(obj.use);
+    g(obj.use);
+}
+```
+
+If every invocation of `.use` or `use ||` here resulted in a call to
+`Clone::clone`, this program would call `Clone::clone` 10 times (and have 11
+`Arc`s to drop); however, the compiler can elide all the uses in `main` and
```

Comment:
`f` takes the `Arc` by-value... how can the compiler elide a `use` here?

Reply:
I think the idea is that since the body of `f` is known and would also "work" with `&Arc`, the compiler could "rewrite" it to instead take `&Arc` using the "If `x` is not statically known to be dead, but *all* of the following conditions are met, the compiler *may* elide an `x.use` and use `&x` instead: ..." optimization listed above.